### PR TITLE
Optimize revocation cache

### PR DESCRIPTION
**Pull request by jeffsantarossa**

### CODE REVIEW

**Comment by hjacobs** [comment]
@jeffsantarossa could you summarize your changes (and especially the "why")?

**Comment by jeffsantarossa** [comment]
To get a list of claim names in the cache and the latest timestamp, we were cycling through the whole cache which is extremely inefficient for large caches. I updated the cache to include a variable which would hold the last pull timestamp and a map that would hold a list of unique claim names.

Running some basic load tests, the tokens validated / min has increased by at least 20x for large caches (> 1.000.000). See attached for latest test result data.
[planb-tokeninfo-loadtest1.txt](https://github.com/zalando/planb-tokeninfo/files/252226/planb-tokeninfo-loadtest1.txt)

**Comment by lmineiro** [comment]
Can you add some Benchmark code next to the tests? I see that you actually went through the process of synthesizing some requests for this. It would be valuable to have them as Benchmark tests in the source tree.

**Comment by lmineiro** [comment]
@jeffsantarossa any plans on providing the benchmarks? that would be really great

**Comment by jeffsantarossa** [comment]
@lmineiro Yesterday was my first day back. I'm just finishing them off.

**Comment by jeffsantarossa** [comment]
Added two types of benchmarks for checking if a token is revoked. One uses a fixed number of claim names and varies the number of revocations in the cache; the other uses a fixed number of revocations and changes the number of claim names stored. 

BenchmarkIsJWTRevokedClaimNames5-4 	   50000	     26605 ns/op
BenchmarkIsJWTRevokedClaimNames10-4	   50000	     47672 ns/op
BenchmarkIsJWTRevokedClaimNames15-4	   30000	     55018 ns/op
BenchmarkIsJWTRevokedClaimNames20-4	   20000	     72068 ns/op


BenchmarkIsJWTRevokedRevocations10K-4 	  100000	     21875 ns/op
BenchmarkIsJWTRevokedRevocations100K-4	  100000	     21752 ns/op
BenchmarkIsJWTRevokedRevocations1M-4  	  100000	     19823 ns/op
BenchmarkIsJWTRevokedRevocations2M-4  	  100000	     19885 ns/op
BenchmarkIsJWTRevokedRevocations3M-4  	  100000	     20110 ns/op
BenchmarkIsJWTRevokedRevocations4M-4	  100000	     19803 ns/op
BenchmarkIsJWTRevokedRevocations5M-4	  100000	     19887 ns/op

I can add benchmarks for expiring tokens in cache as that's the most time consuming operation. Thoughts?

**Comment by hjacobs** [comment]
@vetinari what's the status here? Can you review? @lmineiro could you check, too?

**Comment by hjacobs** [comment]
Guys, how to get this merged?

**Comment by vetinari** [comment]
:+1:

**Comment by jeffsantarossa** [comment]
@lmineiro Do you want to take a look at this before I merge?

**Comment by lmineiro** [comment]
lgtm
